File: backend-examples/php/preact-iso-url-pattern.php
<?php
// Run program: php preact-iso-url-pattern.php

function preactIsoUrlPatternMatch($url, $route) {
    $url = array_filter(explode('/', $url));
    $route = array_filter(explode('/', $route ?? ''));

    for ($i = 0; $i < max(count($url), count($route)); $i++) {
        preg_match('/^(:?)(.*?)([+*?]?)$/', $route[$i] ?? '', $matches);
        $m = $matches[1] ?? '';
        $param = $matches[2] ?? '';
        $flag = $matches[3] ?? '';
        $val = $url[$i] ?? null;

        // segment match:
        if (!$m && $param == $val) continue;

        // /foo/* match
        if (!$m && $val && $flag == '*') {
            break;
        }

        // segment mismatch / missing required field:
        if (!$m || (!$val && $flag != '?' && $flag != '*')) return false;

        $rest = $flag == '+' || $flag == '*';

        if ($rest) break;
    }

    return true;
}

// Example usage:
// var_dump(preactIsoUrlPatternMatch("/foo/bar", "/foo/:param"));
// var_dump(preactIsoUrlPatternMatch("/foo/bar/baz", "/foo/*"));
// var_dump(preactIsoUrlPatternMatch("/foo", "/foo/:param?"));
// var_dump(preactIsoUrlPatternMatch("/foo/bar", "/bar/:param"));
// var_dump(preactIsoUrlPatternMatch('/users/test%40example.com/posts', '/users/:userId/posts'));
?>

File: backend-examples/ruby/preact-iso-url-pattern.rb
# Run program: ruby preact-iso-url-pattern.rb

def preact_iso_url_pattern_match(url, route)
  url = url.split('/').reject(&:empty?)
  route = (route || '').split('/').reject(&:empty?)

  (0...[url.length, route.length].max).each do |i|
    m, param, flag = route[i]&.match(/^(:?)(.*?)([+*?]?)$/)&.captures || ['', '', '']
    val = url[i]

    # segment match:
    next if m.empty? && param == val

    # /foo/* match
    break if m.empty? && val && flag == '*'

    # segment mismatch / missing required field:
    return false if m.empty? || (!val && flag != '?' && flag != '*')

    rest = flag == '+' || flag == '*'

    break if rest
  end

  true
end

# Example usage:
# puts preact_iso_url_pattern_match("/foo/bar", "/foo/:param")
# puts preact_iso_url_pattern_match("/foo/bar/baz", "/foo/*")
# puts preact_iso_url_pattern_match("/foo", "/foo/:param?")
# puts preact_iso_url_pattern_match("/foo/bar", "/bar/:param")
# puts preact_iso_url_pattern_match('/users/test%40example.com/posts', '/users/:userId/posts')